### PR TITLE
Upgraded from future 0.1 to future 0.3

### PR DESCRIPTION
**Pull request by jubeless**
- Modified blockstream implementation to support future 0.3
- Modified cancellable to support future 0.3
- Modified `retry` lib to support future 0.3. Now retry func expects a future 0.3 function
- Sprinkled some `boxed` & `compat` where needed
- Fixed examples

### CODE REVIEW

File: chain/ethereum/src/ethereum_adapter.rs [hunk 1]
```diff
@@ -160,14 +160,14 @@ impl EthereumAdapter {
         }
     }
 
-    fn traces(
-        &self,
-        logger: &Logger,
+    async fn traces(
+        self,
```
Comment (maoueh) [inline chain/ethereum/src/ethereum_adapter.rs:164]: Does this really need to be owned and then dropped? Seems like we should be able to keep `&self` here.

File: core/src/link_resolver.rs [hunk 2]
```diff
@@ -250,10 +246,10 @@ impl LinkResolverTrait for LinkResolver {
                     }
                     Result::<Vec<u8>, reqwest::Error>::Ok(data)
                 }
-                .boxed()
-                .compat()
+                // .boxed()
```
Comment (maoueh) [inline core/src/link_resolver.rs:249]: Dead code

File: graph/src/blockchain/block_stream.rs [hunk 3]
```diff
@@ -651,110 +647,103 @@ impl<C: Blockchain> Stream for BlockStream<C> {
...
+                            self.state = BlockStreamState::BeginReconciliation;
+                            break Poll::Ready(Some(Ok(BlockStreamEvent::Revert(block))));
                         }
-
-                        Ok(Async::NotReady) => {
+                        Poll::Pending => {
                             // Nothing to change or yield yet.
-                            state = BlockStreamState::Reconciliation(next_blocks_future);
-                            break Ok(Async::NotReady);
+                            // self.state =
```
Comment (maoueh) [inline graph/src/blockchain/block_stream.rs:672]: Dead code

File: graph/src/blockchain/block_stream.rs [hunk 4]
```diff
@@ -651,110 +647,103 @@ impl<C: Blockchain> Stream for BlockStream<C> {
...
+                            //     BlockStreamState::Reconciliation(next_blocks_future);
+                            break Poll::Pending; //todo: is this ok
```
Comment (maoueh) [inline graph/src/blockchain/block_stream.rs:674]: Resolve TODO

File: graph/src/blockchain/block_stream.rs [hunk 5]
```diff
@@ -651,110 +647,103 @@ impl<C: Blockchain> Stream for BlockStream<C> {
...
                             self.consecutive_err_count += 1;
 
                             // Pause before trying again
                             let secs = (5 * self.consecutive_err_count).max(120) as u64;
-                            state = BlockStreamState::RetryAfterDelay(Box::new(
-                                tokio::time::delay_for(Duration::from_secs(secs))
-                                    .map(Ok)
-                                    .boxed()
-                                    .compat(),
+                            // (Pin<Box<dyn Future<Output = Result<(), Error>> + Send>>),
```
Comment (maoueh) [inline graph/src/blockchain/block_stream.rs:684]: Dead code

File: graph/src/ext/futures.rs [hunk 6]
```diff
@@ -252,11 +254,16 @@ impl<F: Future> FutureExtension for F {
...
     }
 
-    fn timeout(self, dur: Duration) -> tokio::time::Timeout<Compat01As03<Self>> {
-        tokio::time::timeout(dur, self.compat())
+    fn timeout(self, dur: Duration) -> tokio::time::Timeout<Self> {
+        tokio::time::timeout(dur, self)
     }
 }
 
+pub enum CancelableItem<T> {
```
Comment (maoueh) [inline graph/src/ext/futures.rs:262]: Dead code

File: graph/src/util/futures.rs [hunk 7]
```diff
@@ -9,6 +8,7 @@ use std::time::Duration;
 use thiserror::Error;
 use tokio_retry::strategy::{jitter, ExponentialBackoff};
 use tokio_retry::Retry;
+// use std::future::Future;
```
Comment (maoueh) [inline graph/src/util/futures.rs:11]: Dead code

**Review by maoueh — COMMENTED** [review]
Some quick review about dead code and other small stuff that needed discussion

**Comment by leoyvens** [comment]
@jubeless Did you test this PR by manually running graph-node and checking if the retries work fine? The integration tests are passing which is a good sign, but I'm surprised that this works considering that tokio retry 0.3 depends on tokio 1.0, and mixing tokio versions frequently fails at runtime.

**Comment by otaviopace** [comment]
Besides what @leoyvens pointed out, the CI is breaking because of formatting issues.

You should run `cargo fmt` so it will fix it 🙂 

If for some reason you don't have it installed, take a look [here](https://github.com/rust-lang/rustfmt)

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (otaviopace) [inline CONTRIBUTING.md:1]: Why was this line added?

File: graphql/tests/query.rs [hunk 9]
```diff
@@ -1320,11 +1320,10 @@ fn subscription_gets_result_even_without_events() {
         let results: Vec<_> = stream
             .take(1)
             .collect()
-            .map(Result::<_, ()>::Ok)
-            .compat()
+            // .map(Result::<_, ()>::Ok)
             .timeout(Duration::from_secs(3))
             .await
-            .unwrap()
+            // .unwrap()
```
Comment (otaviopace) [inline graphql/tests/query.rs:1326]: These two comments should be deleted.

**Comment by jubeless** [comment]
Closing this PR an re-open it once we have figured out the outcome of #2664
